### ISSUE TITLE
Templating Proposal

### ISSUE THREAD

**Issue by imaustink** [canjs/canjs#3197]
### TLDR
Create an alternative to `can-stache` that can be used as a drop-in replacement that would allow the use of `Pug` to simplify writing templates and help you write less code.

### Summary
Since I first tried CanJS, my biggest complaint has been the somewhat cumbersome nature of mustache templates. In short, way too much typing. A few colleagues from my previous employer noted the same thing when trying it out. Although we may be a minority, we all thoroughly enjoyed working with the [Pug](https://pugjs.org/api/getting-started.html) (formerly Jade) templating engine.

### About Pug
Pug allows you to type a lot less. For example:
```jade
#container
	ul.someclass
		li Item A
		li Item B
		li Item C
```
- 60 characters including whitespace.

Transpiles to:
```html
<div id="container">
	<ul class="someclass">
		<li>Item A</li>
		<li>Item B</li>
		<li>Item C</li>
	</ul>
</div>
```
- 112 characters including whitespace.

Pug also has features for conditionals, iteration, interpolation, modules, escaping, and much more. It would theoretically support everything needed for CanJS. I am a little fuzzy on how binding would work still, I've had some interesting conversations about that but I think some experimentation is in order.

### Pros
- Less typing
- Simpler syntax
- Arguably more readable

### Cons
- Whitespace sensitivity

### Questions
- How feasible is this all, especially with regards to binding?
- How many people would be interested in using Pug over stache?

**Comment by justinbmeyer** [comment]
What would the binding syntaxes look like?

**Comment by rjgotten** [comment]
My $0,02 :

> Arguably more readable

It's arguably far **_less_** readable. Pug presents a near-complete elimination of context that users need. The loss of HTML-like structured markup means there is a far greater gap between the template and the actual rendered result. This hampers users that use the output HTML as their mental model for reasoning about their templates' structure. 


> Simpler syntax

Short ≠ simple. Overly terse syntax introduces complications in the semantics underlying the syntax and invites ambiguity.


> Less typing

Use an editor with autocompletion and solve that problem at the IDE level.

**Comment by jbsulli** [comment]
I'm new to donejs but saw this and wanted to throw my thoughts out there as a Pug.js lover (hope you guys don't mind):

What Pug brings to the table for me:

**1) Uses JavaScript expressions for binding**
Binding syntax will be more familiar to JavaScript devs.

**2) Consistency when it comes to formatting and indentation**
Quickly jump into other people's templates without beautifying.

**3) No closing tags**
So no more missing or mismatched closing tags.

**4) All the important bits are emphasized**
Ids, classes, conditionals, etc are easy to pick out and digest with a quick glance.

**5) Comments!!! So easy!**

Markup taken from the [donejs in-depth guide](https://donejs.com/place-my-order.html#write-the-template):
```handlebars
<can-import from="can-stache/helpers/route" />

<div class="restaurants">
  <h2 class="page-header">Restaurants</h2>
  <form class="form">
    <div class="form-group">
      <label>State</label>
      <select {($value)}="state" {{#if states.isPending}}disabled{{/if}}>
        {{#if states.isPending}}
          <option value="">Loading...</option>
        {{else}}
          {{^if state}}
          <option value="">Choose a state</option>
          {{/if}}
          {{#each states.value}}
          <option value="{{short}}">{{name}}</option>
          {{/each}}
        {{/if}}
      </select>
    </div>
    <div class="form-group">
      <label>City</label>
      <select {($value)}="city"{{^if state}}disabled{{/if}}>
        {{#if cities.isPending}}
          <option value="">Loading...</option>
        {{else}}
          {{^if city}}
          <option value="">Choose a city</option>
          {{/if}}
          {{#each cities.value}}
          <option value="{{name}}">{{name}}</option>
          {{/each}}
        {{/if}}
      </select>
    </div>
  </form>

  ...
```

Would look something like this in Pug:
```jade
//- 
  What is going on here? 
  Let me tell you using this multi-line comment!
  Note: the dash means this comment is removed when rendered otherwise you get an HTML comment
can-import can-stache/helpers/route

.restaurants
  h2.page-header Restaurants

  //- state and city selection
  form.form
    //- state select
    .form-group
      label State
      //- wait for states to be loaded
      select(value=state disabled=states.isPending)
        if states.isPending
          option(value="") Loading...
        else
          if !state
            option(value="") Choose a state
          //- I may be misreading what is going on in donejs here...
          each name, short in states
            option(value=short)= name

    //- city select
    .form-group
      label City
      //- wait for a state to be selected
      select(value=city disabled=!state)
        if cities.isPending
          option(value="") Loading...
        else
          if !city
            option(value="") Choose a city
          //- I may be misreading what is going on in donejs here...
          each name in cities
            option(value=name) name
```

I can agree with @rjgotten that there is a learning curve and source doesn't look like output but it is my understanding that this would be an alternative to stache and not a replacement?

**Closing thoughts:** It looks popular according to [js.libhunt.com](https://js.libhunt.com/categories/13-templating-engines) and [slant.co](https://www.slant.co/topics/51/~best-javascript-templating-engines) and some positive things to say from [colorlib](https://colorlib.com/wp/top-templating-engines-for-javascript/).

**Comment by rjgotten** [comment]
> 3) No closing tags
So no more missing or mismatched closing tags.

No closing tags, so no way to visually track relative indent level.

The example you've posted already shows this problem in action. As a reader skimming the template, you have to actively put in effort to visually correlate the indent position between the first and second `.form-group`, whereas this problem does not exist in the more HTML-like stache template. The explicit closing tags allow users to easily pop a given number of nesting levels off of the stack they mentally keep track of.
